src/main/java/org/nuxeo/ecm/platform/pdf/service/PDFTransformationServiceImpl.java

```diff
@@ -0,0 +1,247 @@
+            int[] rgb = PDFUtils.hex255ToRGB(properties.getHex255Color());
+
+            for (Object o : pdfDoc.getDocumentCatalog().getAllPages()) {
+                PDPage page = (PDPage) o;
+                PDRectangle pageSize = page.findMediaBox();
+                PDResources resources = page.findResources();
+
+                // Get the defined graphic states.
+                HashMap<String, PDExtendedGraphicsState> graphicsStateDictionary =
+                        (HashMap<String, PDExtendedGraphicsState>) resources.getGraphicsStates();
+                if (graphicsStateDictionary != null) {
+                    graphicsStateDictionary.put("TransparentState",
+                            extendedGraphicsState);
+                    resources.setGraphicsStates(graphicsStateDictionary);
+                } else {
+                    Map<String, PDExtendedGraphicsState> m = new HashMap<>();
+                    m.put("TransparentState", extendedGraphicsState);
+                    resources.setGraphicsStates(m);
+                }
+
+                try (PDPageContentStream contentStream =
+                             new PDPageContentStream(pdfDoc, page, true, true, true)) {
+                    contentStream.beginText();
+                    contentStream.setFont(font, (float) properties.getFontSize());
+                    contentStream.appendRawCommands("/TransparentState gs\n");
+                    contentStream.setNonStrokingColor(rgb[0], rgb[1], rgb[2]);
+                    Point2D position = computeTranslationVector(
+                            pageSize.getWidth(),watermarkWidth,
+                            pageSize.getHeight(),properties.getFontSize(),properties);
+                    contentStream.setTextRotation(
+                            Math.toRadians(properties.getRotation()),
+                            position.getX(),
+                            position.getY());
+                    contentStream.drawString(text);
+                    contentStream.endText();
+                }
+            }
+            return saveInTempFile(pdfDoc);
+
+        } catch (Exception e) {
+            throw new NuxeoException(e);
+        }
+    }
+
+    @Override
+    public Blob applyImageWatermark(Blob input, Blob watermark, WatermarkProperties properties) {
+
+        // Set up the graphic state to handle transparency
+        // Define a new extended graphic state
+        PDExtendedGraphicsState extendedGraphicsState = new PDExtendedGraphicsState();
+        // Set the transparency/opacity
+        extendedGraphicsState.setNonStrokingAlphaConstant((float) properties.getAlphaColor());
+
+        try (PDDocument pdfDoc = PDDocument.load(input.getStream())){
+            BufferedImage image = ImageIO.read(watermark.getStream());
+            PDXObjectImage ximage = new PDPixelMap(pdfDoc, image);
+
+            for (Object o : pdfDoc.getDocumentCatalog().getAllPages()) {
+                PDPage page = (PDPage) o;
```

Comment:
You can probably cast `getAllPages` to `List<PDPage>` in the `for` statement to avoid having an intermediate `Object`.